**Household Power Consumption**

**Dataset Description:**
This dataset captures electric power consumption in a household with one-minute intervals over nearly four years (December 2006 to November 2010). It provides detailed time-series data on various electrical quantities and energy sub-metering.

**Characteristics:**

**Type:** Multivariate, Time-Series

**Tasks:** Regression, Clustering

**Records:**2,075,259

**Period:** 47 months

**Sampling Rate:** One minute

**Attributes:**

**Date:** Observation date (dd/mm/yyyy)

**Time:** Observation time (hh:mm:ss)

**Global_active_power:** Active power consumption (kW)

**Global_reactive_power:** Reactive power consumption (kW)

**Voltage:** Supply voltage (V)

**Global_intensity:** Current intensity (A)

**Notes:**

**Energy Calculation: **(global_active_power*1000/60 - sub_metering_1 - sub_metering_2 - sub_metering_3) gives unmeasured energy consumption. Missing Values: Around 1.25% of rows have missing data.

**Summary**

The dataset is ideal for analyzing energy consumption trends, time-series forecasting, and predictive modeling, with some data cleaning needed due t


**Import libraries and Load the dataset**

In [3]:
import pandas as pd
import numpy as np

In [4]:
!gdown --fuzzy https://drive.google.com/file/d/1bvaXJJqNObOCkX-i475BNxpidk024pyx/view?usp=sharing


Downloading...
From (original): https://drive.google.com/uc?id=1bvaXJJqNObOCkX-i475BNxpidk024pyx
From (redirected): https://drive.google.com/uc?id=1bvaXJJqNObOCkX-i475BNxpidk024pyx&confirm=t&uuid=bcce157f-e535-4b30-bd85-0a085e4f3c3d
To: /content/household_power_consumption.txt
100% 133M/133M [00:03<00:00, 37.7MB/s]


In [5]:
df=pd.read_csv('/content/household_power_consumption.txt',sep=";")


<ipython-input-5-e38478b95906>:1: DtypeWarning: Columns (2,3,4,5,6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('/content/household_power_consumption.txt',sep=";")


**Basic Data Exploration**

In [6]:


df.shape

(2075259, 9)

In [7]:


df.head()

Date      Time Global_active_power Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00               4.216                 0.418  234.840   
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0           18.400          0.000          1.000            17.0  
1           23.000          0.000          1.000            16.0  
2           23.000          0.000          2.000            17.0  
3           23.000          0.000          1.000            17.0  
4           15.800          0.000          1.000            17.0

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Date                   object 
 1   Time                   object 
 2   Global_active_power    object 
 3   Global_reactive_power  object 
 4   Voltage                object 
 5   Global_intensity       object 
 6   Sub_metering_1         object 
 7   Sub_metering_2         object 
 8   Sub_metering_3         float64
dtypes: float64(1), object(8)
memory usage: 142.5+ MB


In [9]:
df.tail()

Date      Time Global_active_power Global_reactive_power  \
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

        Voltage Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
2075254  240.43              4.0            0.0            0.0             0.0  
2075255   240.0              4.0            0.0            0.0             0.0  
2075256  239.82              3.8            0.0            0.0             0.0  
2075257   239.7              3.8            0.0            0.0             0.0  
2075258  239.55              3.8            0.0            0.0             0.0

In [10]:
df.describe()

Sub_metering_3
count    2.049280e+06
mean     6.458447e+00
std      8.437154e+00
min      0.000000e+00
25%      0.000000e+00
50%      1.000000e+00
75%      1.700000e+01
max      3.100000e+01

**Observations:**

**Data Types Issue:** Eight columns are object types instead of numeric, indicating non-numeric values or formatting issues.

**Memory Usage:** The dataset uses 142.5 MB, suggesting a large dataset requiring efficient handling.

**Data Cleaning Needed:** Conversion to numeric types is necessary for meaningful analysis and modeling.

In [11]:


len(df)

2075259

**Observations :**

**Count: **~2,049,280 entries (indicating missing values).

**Mean: **6.46 watt-hours.

**Standard Deviation:** 8.44 (high variability).

**Minimum:** 0 watt-hours (no consumption).

**Percentiles:**

25% of values are 0 watt-hours.

50% (median) is 1 watt-hour.

75% are up to 17 watt-hours.

**Maximum:** 31 watt-hours.

Overall, the data shows many instances of low or zero energy consumption.

**Datatype conversion**

In [12]:
# Convert all columns to float, coercing errors
for column in df.columns:
    df[column] = pd.to_numeric(df[column], errors='coerce')

In [13]:
# Display DataFrame information after conversion
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Date                   float64
 1   Time                   float64
 2   Global_active_power    float64
 3   Global_reactive_power  float64
 4   Voltage                float64
 5   Global_intensity       float64
 6   Sub_metering_1         float64
 7   Sub_metering_2         float64
 8   Sub_metering_3         float64
dtypes: float64(9)
memory usage: 142.5 MB


**Observations**

**Data Type Conversion:** All columns in the DataFrame are now of type float64, indicating successful conversion. Any non-numeric entries were coerced to NaN.

**Columns Overview: **The DataFrame contains 9 columns, including Date, Time, Global_active_power, Global_reactive_power, Voltage, Global_intensity, Sub_metering_1, Sub_metering_2, and Sub_metering_3.

**Missing Values:** bold text The conversion process may have introduced NaN values in the Date and Time columns, which need to be addressed for accurate analysis.

**Memory Usage:** The DataFrame occupies approximately 156.3 MB of memory with 2,049,280 entries.

In [14]:
# Display descriptive statistics of a DataFrame after convertion of columns to float
df.describe()


Date  Time  Global_active_power  Global_reactive_power       Voltage  \
count   0.0   0.0         2.049280e+06           2.049280e+06  2.049280e+06   
mean    NaN   NaN         1.091615e+00           1.237145e-01  2.408399e+02   
std     NaN   NaN         1.057294e+00           1.127220e-01  3.239987e+00   
min     NaN   NaN         7.600000e-02           0.000000e+00  2.232000e+02   
25%     NaN   NaN         3.080000e-01           4.800000e-02  2.389900e+02   
50%     NaN   NaN         6.020000e-01           1.000000e-01  2.410100e+02   
75%     NaN   NaN         1.528000e+00           1.940000e-01  2.428900e+02   
max     NaN   NaN         1.112200e+01           1.390000e+00  2.541500e+02   

       Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
count      2.049280e+06    2.049280e+06    2.049280e+06    2.049280e+06  
mean       4.627759e+00    1.121923e+00    1.298520e+00    6.458447e+00  
std        4.444396e+00    6.153031e+00    5.822026e+00    8.437154e+00  
min        2.000000e-01    0.000000e+00    0.000000e+00    0.000000e+00  
25%        1.400000e+00    0.000000e+00    0.000000e+00    0.000000e+00  
50%        2.600000e+00    0.000000e+00    0.000000e+00    1.000000e+00  
75%        6.400000e+00    0.000000e+00    1.000000e+00    1.700000e+01  
max        4.840000e+01    8.800000e+01    8.000000e+01    3.100000e+01

**Observations**

**Count:** All numeric columns have 2,049,280 entries, but Date and Time have zero counts, indicating they were converted to NaN.

**Mean Values:**

**Global_active_power:** 1.09

**Global_reactive_power:** 0.12

**Voltage:** 240.84

**Global_intensity:** 4.63


**Range:**

**Global_active_power:** 0.076 to 11.12

Conclusion The dataset needs handling for NaN values in Date and Time columns, and further analysis is required for numeric features.

**Finding Null Values and Unique Values**

In [17]:
# Exclude date and time columns by their names
columns_to_exclude = ['Date', 'Time']

# Select all columns except the ones in 'columns_to_exclude'
df_filtered = df.drop(columns=columns_to_exclude)

# Now get unique counts for the remaining columns
unique_values = df_filtered.nunique()
print(unique_values)

Global_active_power      4186
Global_reactive_power     532
Voltage                  2837
Global_intensity          221
Sub_metering_1             88
Sub_metering_2             81
Sub_metering_3             32
dtype: int64


In [21]:
# Count of unique values in each column
unique_values = df_filtered.nunique()

print(unique_values)

Global_active_power      4186
Global_reactive_power     532
Voltage                  2837
Global_intensity          221
Sub_metering_1             88
Sub_metering_2             81
Sub_metering_3             32
dtype: int64


In [25]:
columns_to_exclude = ['Date', 'Time']

# Drop those columns and check for missing values in the remaining columns
missing_values = df.drop(columns=columns_to_exclude).isnull().any()
print(missing_values)

Global_active_power      True
Global_reactive_power    True
Voltage                  True
Global_intensity         True
Sub_metering_1           True
Sub_metering_2           True
Sub_metering_3           True
dtype: bool


In [28]:
columns_to_exclude = ['Date', 'Time']

# Drop those columns and count missing values in the remaining columns
null_values = df.drop(columns=columns_to_exclude).isnull().sum()
print(null_values)

Global_active_power      25979
Global_reactive_power    25979
Voltage                  25979
Global_intensity         25979
Sub_metering_1           25979
Sub_metering_2           25979
Sub_metering_3           25979
dtype: int64


**Handling Null Values**

In [29]:

# Drop rows where Sub_metering_3 is null
df = df.dropna(subset=['Sub_metering_3'])
df


Date  Time  Global_active_power  Global_reactive_power  Voltage  \
0         NaN   NaN                4.216                  0.418   234.84   
1         NaN   NaN                5.360                  0.436   233.63   
2         NaN   NaN                5.374                  0.498   233.29   
3         NaN   NaN                5.388                  0.502   233.74   
4         NaN   NaN                3.666                  0.528   235.68   
...       ...   ...                  ...                    ...      ...   
2075254   NaN   NaN                0.946                  0.000   240.43   
2075255   NaN   NaN                0.944                  0.000   240.00   
2075256   NaN   NaN                0.938                  0.000   239.82   
2075257   NaN   NaN                0.934                  0.000   239.70   
2075258   NaN   NaN                0.932                  0.000   239.55   

         Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
0                    18.4             0.0             1.0            17.0  
1                    23.0             0.0             1.0            16.0  
2                    23.0             0.0             2.0            17.0  
3                    23.0             0.0             1.0            17.0  
4                    15.8             0.0             1.0            17.0  
...                   ...             ...             ...             ...  
2075254               4.0             0.0             0.0             0.0  
2075255               4.0             0.0             0.0             0.0  
2075256               3.8             0.0             0.0             0.0  
2075257               3.8             0.0             0.0             0.0  
2075258               3.8             0.0             0.0             0.0  

[2049280 rows x 9 columns]

In [ ]:
# Fill missing values in Sub_metering_3 with the median and assign it back
df['Sub_metering_3'] = df['Sub_metering_3'].fillna(df['Sub_metering_3'].median())


In [30]:
# Display the updated DataFrame (for the first few rows)
print("Updated DataFrame:")
print(df.head())

Updated DataFrame:
   Date  Time  Global_active_power  Global_reactive_power  Voltage  \
0   NaN   NaN                4.216                  0.418   234.84   
1   NaN   NaN                5.360                  0.436   233.63   
2   NaN   NaN                5.374                  0.498   233.29   
3   NaN   NaN                5.388                  0.502   233.74   
4   NaN   NaN                3.666                  0.528   235.68   

   Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
0              18.4             0.0             1.0            17.0  
1              23.0             0.0             1.0            16.0  
2              23.0             0.0             2.0            17.0  
3              23.0             0.0             1.0            17.0  
4              15.8             0.0             1.0            17.0  


In [32]:
# Specify the columns to exclude (Date and Time columns)
columns_to_exclude = ['Date', 'Time']

# Drop the Date and Time columns and fill missing values in the remaining columns
df_filtered = df.drop(columns=columns_to_exclude).fillna(0)

# Count the null values after filling
null_values = df_filtered.isnull().sum()
print("\nNull Values Count After Filling (Excluding Date and Time Columns):")
print(null_values)


Null Values Count After Filling (Excluding Date and Time Columns):
Global_active_power      0
Global_reactive_power    0
Voltage                  0
Global_intensity         0
Sub_metering_1           0
Sub_metering_2           0
Sub_metering_3           0
dtype: int64
